Firestore/core/src/firebase/firestore/timestamp.cc

```diff
@@ -29,46 +29,73 @@
 
 namespace firebase {
 
-Timestamp::Timestamp() {
+namespace {
+
+// We pass it by value without std::move.
+static_assert(std::is_trivially_copyable<Timestamp>::value,
+              "Timestamp must be trivially copyable");
+
+constexpr int32_t kNanosPerSecond = 1E9;
+
+/**
+ * Creates a `Timestamp` from the given non-normalized inputs.
+ *
+ * Timestamp protos require `Timestamp` to always has a positive number of
+ * nanoseconds that is counting forward. For negative time, we need to adjust
+ * representations with negative nanoseconds. That is make (negative seconds s1
+ * + negative nanoseconds ns1) to (negative seconds s2 + positive nanoseconds
+ * ns2). Since nanosecond part is always less than 1 second in our
+ * representation, instead of starting at s1 and going back ns1 nanoseconds,
+ * start at (s1 minus one second) and go *forward* ns2 = (1 second + ns1, ns1 <
+ * 0) nanoseconds.
+ */
+Timestamp MakeNormalizedTimestamp(int64_t seconds, int64_t nanos) {
+  if (nanos < 0) {
+    // Note: if nanoseconds are negative, it must mean that seconds are
+    // non-positive, but the formula would still be valid, so no need to check.
+    seconds = seconds - 1;
+    nanos = kNanosPerSecond + nanos;
+  }
+
+  HARD_ASSERT(nanos < kNanosPerSecond);
+
+  return {seconds, static_cast<int32_t>(nanos)};
 }
 
+}  // namespace
+
 Timestamp::Timestamp(const int64_t seconds, const int32_t nanoseconds)
     : seconds_(seconds), nanoseconds_(nanoseconds) {
   ValidateBounds();
 }
 
 Timestamp Timestamp::Now() {
 #if defined(__APPLE__)
-  // Originally, FIRTimestamp used NSDate to get current time. This method is
+  // Originally, FIRTimestamp used NSDate to get current time. This method
   // preserves the accuracy of that method.
```

Comment:
Very optional: maybe clarify that the old method had _lower_ accuracy which is what we're preserving? Just adding `lower` in front of `accuracy` is enough for me.

Reply:
Done.